platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/buildtree/control/BuildModelParametersProvider.kt

```diff
@@ -0,0 +1,255 @@
+
+    @JvmStatic
+    val isolatedProjectsParallel =
+        InvocationScenarioParameter.Option("org.gradle.internal.isolated-projects.parallel", InvocationScenarioParameter.ANY)
+
+    @JvmStatic
+    val isolatedProjectsCaching =
+        InvocationScenarioParameter.Option("org.gradle.internal.isolated-projects.caching", InvocationScenarioParameter.NONE)
+
+    private
+    val resilientModelBuilding =
+        InternalFlag("org.gradle.internal.resilient-model-building", false)
+
+    /**
+     * Determines Gradle features and behaviors that are required or requested by the build action.
+     *
+     * This includes features like Configuration Cache, Isolated Projects
+     * and their sub-behaviors (e.g., caching and parallelism controls).
+     *
+     * @throws org.gradle.api.GradleException if the requirements are contradictory
+     */
+    @JvmStatic
+    fun parameters(requirements: BuildActionModelRequirements): BuildModelParameters {
+        val startParameter = requirements.startParameter
+        val options = DefaultInternalOptions(startParameter.systemPropertiesArgs)
+        warnOnPreviouslyExistingOptions(options)
+
+        val ccDisabledReason = getConfigurationCacheDisabledReason(startParameter)
+        return when {
+            // TODO:isolated should this also disable IP?
+            ccDisabledReason != null -> vintageMode(requirements, startParameter, options, ccDisabledReason)
+            startParameter.isolatedProjects.get() -> isolatedProjectsMode(requirements, startParameter, options)
+            startParameter.configurationCache.get() ->
+                if (requirements.isCreatesModel)
+                    vintageMode(requirements, startParameter, options)
+                else
+                    configurationCacheTasksOnlyMode(requirements, startParameter, options)
+
+            else -> vintageMode(requirements, startParameter, options)
+        }
+    }
+
+    /**
+     * Compute parameters for nested build trees, which are created by the `GradleBuild` tasks.
+     *
+     * Many features are not supported for nested build trees, such as Configuration Cache or Isolated Projects.
+     *
+     * @see org.gradle.internal.build.NestedRootBuildRunner
+     */
+    @JvmStatic
+    fun parametersForNestedBuildTree(startParameter: StartParameterInternal): BuildModelParameters {
+        return GradleVintageMode(
+            requiresToolingModels = true,
+            parallelProjectExecution = startParameter.isParallelProjectExecutionEnabled,
+            configureOnDemand = startParameter.isConfigureOnDemand,
+            configurationCacheDisabledReason = null,
+            parallelToolingActions = false,
+            resilientModelBuilding = false
+        )
+    }
+
+    private
+    fun vintageMode(
+        requirements: BuildActionModelRequirements,
+        startParameter: StartParameterInternal,
+        options: InternalOptions,
+        ccDisabledReason: String? = null
+    ): GradleVintageMode {
```

Comment:
💅 

I'd go with the interface type here, same for the other methods for CC and IP. AFAIU, The mode classes are just an internal implementation decision on how to implement `BuildModelParameters`